processing/src/test/java/org/apache/druid/segment/IndexBuilder.java

```diff
@@ -267,4 +395,44 @@ private static IncrementalIndex buildIncrementalIndexWithRows(
     }
     return incrementalIndex;
   }
+
+  private static IncrementalIndex buildIncrementalIndexWithInputSource(
+      IncrementalIndexSchema schema,
+      InputSource inputSource,
+      InputFormat inputFormat,
+      @Nullable TransformSpec transformSpec,
+      File inputSourceTmpDir,
+      int maxRows
+  )
+  {
+    Preconditions.checkNotNull(schema, "schema");
+    Preconditions.checkNotNull(inputSource, "inputSource");
+    Preconditions.checkNotNull(inputFormat, "inputFormat");
+    Preconditions.checkNotNull(inputSourceTmpDir, "inputSourceTmpDir");
+
+    final IncrementalIndex incrementalIndex = new OnheapIncrementalIndex.Builder()
+        .setIndexSchema(schema)
+        .setMaxRowCount(maxRows)
+        .build();
+    TransformSpec tranformer = transformSpec != null ? transformSpec : TransformSpec.NONE;
+    InputRowSchema rowSchema = new InputRowSchema(schema.getTimestampSpec(), schema.getDimensionsSpec(), null);
+    InputSourceReader reader = inputSource.reader(rowSchema, inputFormat, inputSourceTmpDir);
+    InputSourceReader transformingReader = tranformer.decorate(reader);
+    try (CloseableIterator<InputRow> rowIterator = transformingReader.read()) {
+      while (rowIterator.hasNext()) {
+        incrementalIndex.add(rowIterator.next());
```

Comment:
When `maxRows` is hit, are we expecting an exception?  Generally speaking, setting the `maxRows` on the tests is done as a way to force running queries against multiple segments, so I had expected to see a check for the numRows and incremental persists in anything that takes `maxRows`.

Reply:
Hmm, good question, this looks wired up to the `maxRowCount` but callers are not checking `canAppendRow` or anything like that and `buildIncrementalIndex` can only return a single `IncrementalIndex` so I think all it can do is explode. There is also `intermediatePersistSize` which can be used to force a bunch of incremental indexes to be written when `mergeIndexes` is called to make sure that segment merging happens.

Nothing much seems to be explicitly setting either of these things, and maybe could be removed or reworked in a follow-up.